rfcs/0096-experimental-nix-derivations.md

```diff
@@ -0,0 +1,81 @@
+---
+feature: experimental-nix-derivations
+start-date: 2021-06-25
+author: gytis-ivaskevicius
+co-authors:
+shepherd-team: edolstra, blaggacao, maljub01, nrdxp
+shepherd-leader: edolstra
+related-issues:
+---
+
+# Summary
+[summary]: #summary
+
+Add a way for users to use experimental Nix features with improved user-experience
+and better communication to the user that they are using an experimental feature.
+
+# Motivation
+[motivation]: #motivation
+
+Motivation:
+- Revert of (this)[https://github.com/NixOS/nixpkgs/pull/123898].
+- Noticed quite a few peeps who were very supportive of [this](https://github.com/NixOS/nixpkgs/pull/120141) PR (which got reverted).
+- Lack of visibility for the end-users that they are using experimental features.
+
+Expected outcome:
+- Improved UX.
+- Reduced risk of users using experimental features in production environments.
+- Reduced non-bug questions/issues on common support channels.
+
+# Detailed design
+[design]: #detailed-design
+
+`nixUnstable` derivation definitions should be updated in such a way so end users would be able to overwrite features with something along the lines of this:
+```nix
+nixUnstable.override { experimentalFeatures = ["flakes"]; }
+```
+
+What's so special about this override:
+- Makes specified experimental features enabled by default. (A little similar to [this](https://github.com/NixOS/nixpkgs/pull/120141) PR)
+- When using these overwritten derivations warning should be shown.
+
+
+Warning expected to look something like this:
+```
+WARNING: You are using experimental Nix features which are up to a subject to change. For more information visit https://nixos.org/xyz
+```
+
+A webpage is expected to contain information as such:
+- Support channels if any.
+- A little rundown of the current state of the feature(s).
+- Some "URL dump" kind of wiki page that that would contain a list of links to issues with existing bugs/workarounds/edge-cases.
```

Comment:
```suggestion
Currently, experimental features are undiscoverable and undocumented. It would be nice if there was a way to query which experimental features are available in a Nix release, and to get some documentation about them (like a link to an associated RFC and possibly links to additional resources)
```